### PR TITLE
FEATURE: tags editing

### PR DESCRIPTION
**Pull request by dimaip**
TODO:
- [ ] editing of tag collections. Could be done later.

**What I did**

Treat tags as entities and support editing them via the inspector.

**How to verify it**

Try to rename some tags.

### CODE REVIEW

**Review by bwaidelich — COMMENTED** [review]
As already posted on Slack: I would strongly recommend not to expose tag ids to the GraphQL API.
That tags aren't value objects but entities with technical id, parent & child tags in the Media package is an architectural flaw IMO. But with this package we have the chance to "iron out" some of these gremlins in the GraphQL API.
Regarding exposing the id:
`deleteTag("dfa06b4b-dc0c-444a-a5a1-7ca2da81c4e3")` is weird compared to `deleteTag("campaign-2020")`. And the fact that I can set the label of two tags to the same value making them indistinguishable in the UI is something I've haver seen with tags anywhere else.

**Comment by dimaip** [comment]
> parent & child tags in the Media package is an architectural flaw IMO

why?

> I can set the label of two tags to the same value making them indistinguishable in the UI is something I've haver seen with tags anywhere else.

I'll add a protection against that, it's quite easy to do. We already disallow creating tags with the same name, we should just prevent it happening via renaming.

> deleteTag("dfa06b4b-dc0c-444a-a5a1-7ca2da81c4e3") is weird compared to deleteTag("campaign-2020").

I'm not opinionated on this, but I'd actually prefer the id variant of this.

It's just much easier to manage tags as entities the way apollo-client works. It was a lot harder to make sense of things with the old implementation and to keep all app state in sync. E.g. when you rename a tag it doesn't suddenly become a new thing from apollo's perspective.
And how do you feel about this @Sebobo?  

Anyways, I'm known for my ugly and too pragmatic solutions, so I'm open to change my mind on this.

**Comment by Sebobo** [comment]
I would love to have a tag as a simple string. But as we allow to edit tags it makes our live harder.
Keeping the UI in sync when modifying a tag is error prone if we don't have an unrelated identifier.

I don't know how this is solved in other software when not having an identifier.

And yes I want to prevent the users from creating 2 collections with the same name and two tags with the same name.

**Comment by bwaidelich** [comment]
>> parent & child tags in the Media package is an architectural flaw IMO

> why?

"Flaw" was maybe a bit harsh but at the very least it's quite uncommon and introduced a lot of challenges we now have to solve (UX and implementation wise). For a hierarchical organization of assets the notion of a "folder" would suit better IMO – Or make the _AssetCollections_ nestable instead.

> It's just much easier to manage tags as entities the way apollo-client works.

For Apollo it should be the same, it can treat the tag scalar as id since it will never change. But even if it gets more cumbersome: I don't think we should design the API solely around one client that uses it if we want to make it usable for other tools (and humans) too

> Anyways, I'm known for my ugly and too pragmatic solutions

Ah, come on, that's flubdub and you know it

> But as we allow to edit tags it makes our live harder.

The value of a Value Object can't be changed, but IMO that's a good thing. We could instead allow to replace a tag for example (if desired we could keep the API as is and just replace the tag under the hood but IMO that should even be explicit).

> I want to prevent the users from creating 2 collections with the same name and two tags with the same name.

For tags that would be solved by making the value its id. For asset collections: If we really allow them to be nested (at some point) we could even extend this and only disallow the same name on one level (like folders work)


Are you up for a quick slack call later today?

**Comment by Sebobo** [comment]
Yes I would have time from 14:30 on.

File: Resources/Private/JavaScript/src/hooks/useSelectedTag.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import { Tag } from '../interfaces';
+import { selectedTagIdState } from '../state';
+import { TAG_FRAGMENT } from '../queries';
+
+const useSelectedTag = (): Tag => {
+    const client = useApolloClient();
+    const selectedTagId = useRecoilValue(selectedTagIdState);
+
+    // Read asset selection from cache as we can only select tags that have been queries before
```
Comment (Sebobo) [inline Resources/Private/JavaScript/src/hooks/useSelectedTag.ts:12]:
```suggestion
    // Read selection from cache as we can only select tags that have been queried before
```

File: Resources/Private/JavaScript/src/hooks/useSelectedTag.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            {
+                id: `Tag_${selectedTagId}`,
+                fragment: TAG_FRAGMENT,
+                fragmentName: 'TagProps'
+            },
+            true
+        );
+    } catch (e) {
+        // TODO: Run query to get the tag when its not found
+        console.error(e, 'selected asset collecion missing in cache');
```
Comment (Sebobo) [inline Resources/Private/JavaScript/src/hooks/useSelectedTag.ts:24]:
```suggestion
        console.error(e, 'Selected tag missing in cache');
```

File: Resources/Private/JavaScript/src/hooks/useUpdateTag.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        action({
+            variables: {
+                id: tag.id,
+                label
+            },
+            optimisticResponse: {
+                updateTag: {
+                    ...tag,
+                    label,
+                    ...(label
```
Comment (Sebobo) [inline Resources/Private/JavaScript/src/hooks/useUpdateTag.ts:29]: Why the label twice?

File: Resources/Private/JavaScript/src/components/SideBarRight/Inspector/TagInspector.tsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        if (label !== selectedTag.label) {
+            updateTag({
+                tag: selectedTag,
+                label
+            })
+                .then(() => {
+                    Notify.ok(translate('actions.updateTag.success', 'The tag has been updated'));
+                })
+                .catch(({ message }) => {
+                    Notify.error(translate('actions.deleteTag.error', 'Error while updating the tag'), message);
```
Comment (Sebobo) [inline Resources/Private/JavaScript/src/components/SideBarRight/Inspector/TagInspector.tsx:43]:
```suggestion
                    Notify.error(translate('actions.updateTag.error', 'Error while updating the tag'), message);
```

File: Resources/Private/JavaScript/src/components/SideBarRight/Inspector/TagInspector.tsx [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                <TextInput type="text" value={label || ''} onChange={setLabel} onEnterKey={handleApply} />
+            </Property>
+
+            <Actions
+                handleApply={handleApply}
+                handleDiscard={handleDiscard}
+                hasUnpublishedChanges={hasUnpublishedChanges}
+            />
+
+            <TagSelectBoxAssetCollection />
```
Comment (Sebobo) [inline Resources/Private/JavaScript/src/components/SideBarRight/Inspector/TagInspector.tsx:67]: Is this the correct component to use here?
We want the list of collections the tag is assigned to correct?

**Comment by Sebobo** [comment]
@dimaip I fixed several errors and did some cleanups. Also made it mergeable again. 
Do you want to take another look yourself?

But we really need more tests when we now add features. Some stuff that was working before was kaputt.

**Comment by Sebobo** [comment]
@dimaip I would also like to get rid of those select box variants for collections and tags. They only wrap the original box with the update action. I would rather put this into the inspector and only have one select box that works for all.

**Pull request merged**